lib/nrf_modem_lib/trace_backends/ram/ram.c

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <sys/errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/ring_buffer.h>
+#include <zephyr/logging/log.h>
+#include <modem/trace_backend.h>
+
+LOG_MODULE_REGISTER(modem_trace_backend, CONFIG_MODEM_TRACE_BACKEND_LOG_LEVEL);
+
+/* don't put this in noinit, memfault won't be able to capture it */
+#define RING_BUF_DECLARE_STATIC(name, size8) \
+	BUILD_ASSERT(size8 < RING_BUFFER_MAX_SIZE,\
+		RING_BUFFER_SIZE_ASSERT_MSG); \
+	static uint8_t _ring_buffer_data_##name[size8]; \
+	static struct ring_buf name = { \
+		.buffer = _ring_buffer_data_##name, \
+		.size = size8 \
+	}
+
+RING_BUF_DECLARE_STATIC(ram_trace_buf, CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_RAM_LENGTH);
+
+static trace_backend_processed_cb trace_processed_callback;
+
+int trace_backend_init(trace_backend_processed_cb trace_processed_cb)
+{
+	trace_processed_callback = trace_processed_cb;
+	return 0;
+}
+
+int trace_backend_deinit(void)
+{
+	/* nothing happens */
+	return 0;
+}
+
+size_t trace_backend_data_size(void)
+{
+	return ring_buf_size_get(&ram_trace_buf);
+}
+
+int trace_backend_read(void *buf, size_t len)
+{
+	return ring_buf_get(&ram_trace_buf, buf, len);
+}
+
+int trace_backend_write(const void *data, size_t len)
+{
+	__ASSERT_NO_MSG(len <= ring_buf_capacity_get(&ram_trace_buf));
```

Comment:
It might be that we should handle this situation by wrapping around.
I'm not convinced assert is the right thing to do if this is considered a normal situation. If possible, we should have a build assert instead, ensuring that max write fragment size fits into the ring buffer, or just wrap around overwriting existing data.